SI-8372: unspliceable type printed in error message

gkossakowski opened the pull request:
The b8a76f6 introduced `ArrayOps.{unzip, unzip3}` methods. Both of those methods have ClassTags as context bounds on their type parameters so they can create (and return) instances of Arrays.
The type inference for those methods is supposed to be guided by implicit evidence that `T <: (T1, T2)` (or `T <: (T1, T2, T3)` in unzip3 case). However, context bounds are desugared into implicit parameters that
prepended in front of implicit parameters declared in source code. That means the implicit evidence won't have a chance to guide type inference because it comes as last implicit parameter.

This commit desugars context bounds and puts them at the end of implicit parameter list. This way type inference is guided properly and we get expected compiler errors for missing class tags.

gkossakowski committed:
Test case for SI-8372: issue with ArrayOps.unzip

This commit merely documents current (suboptimal) error message printed
for the code reported in SI-8372. The next commit will fix the problem.

gkossakowski commented:
Review by @adriaanm.

// cc @Ichoran

xeno-by commented on src/library/scala/collection/mutable/ArrayOps.scala:
```diff
@@ -114,10 +114,14 @@ trait ArrayOps[T] extends Any with ArrayLike[T, Array[T]] with CustomParalleliza
    *                of each element pair of this Array.
    */
-  def unzip[T1: ClassTag, T2: ClassTag](implicit asPair: T => (T1, T2)): (Array[T1], Array[T2]) = {
+  def unzip[T1, T2](implicit asPair: T => (T1, T2), ct1: ClassTag[T1], ct2: ClassTag[T2]): (Array[T1], Array[T2]) = {
```
What's the difference here?

gkossakowski replied:
See the commit message. The order of implicit parameters is different.

xeno-by replied:
Would be cool to have this documented. I mean in comments here.

gkossakowski replied:
The comments above serve as documentation of the method for users. I'm not sure if details of type inference should be part of that. Do you have anything specific that you would like me to write here?

xeno-by replied:
Fair enough. Then how about writing a normal `//` comment right above the unzip signature that would say that the signature can't be collapsed into context bounds, because some brief explanation?

gkossakowski replied:
Done.

xeno-by commented on src/library/scala/collection/mutable/ArrayOps.scala:
```diff
@@ -137,10 +143,19 @@ trait ArrayOps[T] extends Any with ArrayLike[T, Array[T]] with CustomParalleliza
    */
-  def unzip3[T1: ClassTag, T2: ClassTag, T3: ClassTag](implicit asTriple: T => (T1, T2, T3)): (Array[T1], Array[T2], Array[T3]) = {
+  // implementation NOTE: ct1, ct2, ct3 can't be written as context bounds because desugared
+  // implicits are put in front of asPair paremeter that is supposed to guide type inference
```
Typo in "parEmeter" :)

gkossakowski replied:
Fixed. Thanks!

gkossakowski committed:
SI-8372: unspliceable type printed in error message

The b8a76f688c6ce2a4c305da064303bb46b53be875 introduced ArrayOps.{unzip,
unzip3} methods. Both of those methods have ClassTags as context bounds on
their type parameters so they can create (and return) instances of Arrays.
The type inference for those methods is supposed to be guided by
implicit evidence that T <: (T1, T2) (or T <: (T1, T2, T3) in unzip3 case).
However, context bounds are desugared into implicit parameters that
prepended in front of implicit parameters declared in source code. That
means the implicit evidence won't have a chance to guide type inference
because it comes as last implicit parameter.

This commit desugars context bounds and puts them at the end of implicit
parameter list. This way type inference is guided properly and we get
expected compiler errors for missing class tags.

The change to parameters order breaks binary compatibility with respect
to 2.11.0-RC1. I added filters to our binary compatibility configuration
files. We can get rid of them as soon as 2.11.0 is out.

Ichoran commented:
LGTM.  I didn't realize that implicit evidence was so strongly ordered within an implicit parameter list or I'd have done it this way to begin with.  Did the old way actually make valid code not work, or was it just an issue of emitting an unhelpful error message?

adriaanm commented:
yeah, the order of the implicit arguments matters for type inference, it's kind of like fundeps in haskell

Ichoran commented:
And I just checked the other bug report that showed that it did actually require the pair first in order to get correct behavior in some cases.  Fair enough.
